Add human readable reporter that’s also funny because emoji.

justMaku opened the pull request:
This pull request adds a new reporter type that's much easier to read for human eye. I believe it could be useful for people that use swiftlint for example in a pre-commit hook or on CI.

**Example:**
```
/Users/maku/Projects/Temp/Otter.swift
⛔️ Line 159: Function should have complexity 10 or less: currently complexity equals 21
⛔️ Line 185: Function should have complexity 10 or less: currently complexity equals 96
⚠️ Line 12: Lines should not have trailing whitespace.
⚠️ Line 18: Lines should not have trailing whitespace.
⚠️ Line 115: Lines should not have trailing whitespace.

/Users/maku/Projects/Temp/Other.swift
⚠️ Line 23: Lines should not have trailing whitespace.
⚠️ Line 69: Lines should not have trailing whitespace.
⚠️ Line 92: Lines should not have trailing whitespace.
```

jpsim commented on CHANGELOG.md:
```diff
@@ -11,6 +11,8 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#910](https://github.com/realm/SwiftLint/issues/910)
   [#911](https://github.com/realm/SwiftLint/issues/911)
+* Add `EmojiReporter`: a human readble reporter.
```
we generally use a single empty line between changelog entries

jpsim commented on Source/SwiftLintFramework/Extensions/Array+SwiftLint.swift:
```diff
@@ -24,3 +24,26 @@ extension Array where Element: NSTextCheckingResult {
+}
+
+public extension SequenceType {
+    func categorise<U: Hashable>(@noescape by transform: Generator.Element -> U) -> [U:[Generator.Element]] {
```
we use American English spellings throughout the project: `categorize`

jpsim commented on Source/SwiftLintFramework/Extensions/Array+SwiftLint.swift:
```diff
@@ -24,3 +24,26 @@ extension Array where Element: NSTextCheckingResult {
+
+public extension SequenceType {
+    func categorise<U: Hashable>(@noescape by transform: Generator.Element -> U) -> [U:[Generator.Element]] {
+        var dictionary: [U:[Generator.Element]] = [:]
```
we're in the process of adding a new rule requiring a space after the colon in dictionaries: `var dictionary = [U: [Generator.Element]]()`

jpsim commented on Source/SwiftLintFramework/Reporters/EmojiReporter.swift:
```diff
@@ -0,0 +1,70 @@
+
+    public static func generateReport(violations: [StyleViolation]) -> String {
+        return violations.categorise {
+                (violation) in
```
closure parameters on the same line as opening brace please: `{ violation in`

jpsim commented on Source/SwiftLintFramework/Reporters/EmojiReporter.swift:
```diff
@@ -0,0 +1,70 @@
+                (violation) in
+                violation.location.file ?? "Other"
+            }.map {
+                (filename, violations) in
```
closure parameters on the same line as opening brace please: `{ filename, violations in`

justMaku commented:
Thanks for the feedback @jpsim!

I've updated the PR with the changes you've requested and also added a mention about the emoji reporter to the both README files.

jpsim commented on CHANGELOG.md:
```diff
@@ -26,6 +26,9 @@
   [#910](https://github.com/realm/SwiftLint/issues/910)
   [#911](https://github.com/realm/SwiftLint/issues/911)
 
+* Add `EmojiReporter`: a human readble reporter.
```
missing two trailing spaces. See https://github.com/realm/SwiftLint/blob/master/CONTRIBUTING.md#tracking-changes

justMaku replied:
Oh, okay 😅

justMaku replied:
Out of curiosity: what's the reason behind this rule?

jpsim commented on Source/SwiftLintFramework/Extensions/Array+SwiftLint.swift:
```diff
@@ -24,3 +24,27 @@ extension Array where Element: NSTextCheckingResult {
 }
+
+extension Array where Element: Equatable {
+    var unique: [Element] {
```
`internal var`?

justMaku replied:
Isn't it internal by default?

jpsim replied:
yes, but we prefer explicit ACLs.

justMaku replied:
Should I add them everywhere in the file then?

jpsim replied:
nah, it's fine, this file in particular doesn't seem to follow that convention.

jpsim commented on Source/SwiftLintFramework/Extensions/Array+SwiftLint.swift:
```diff
@@ -24,3 +24,27 @@ extension Array where Element: NSTextCheckingResult {
+    }
+}
+
+public extension SequenceType {
```
`internal extension`? This doesn't seem to be relevant in SwiftLint's public API.

justMaku replied:
Right, I copy-pasted it from my other project, doesn't make any sense for this to be public.

jpsim commented on Source/SwiftLintFramework/Extensions/Array+SwiftLint.swift:
```diff
@@ -24,3 +24,27 @@ extension Array where Element: NSTextCheckingResult {
+
+public extension SequenceType {
+    // swiftlint:disable line_length
+    func categorize<U: Hashable>(@noescape by transform: Generator.Element -> U) -> [U: [Generator.Element]] {
```
maybe `group` instead of `categorize`?

justMaku replied:
Hmmm… I was wondering about the same thing, but you're right, `group` sound much nicer.

jpsim replied:
By the way, SourceKitten already has implementations for these, in case you want to compare:

https://github.com/jpsim/SourceKitten/blob/0.15.2/Source/SourceKittenFramework/ClangTranslationUnit.swift#L16-L33

jpsim commented on Source/SwiftLintFramework/Extensions/Array+SwiftLint.swift:
```diff
@@ -24,3 +24,27 @@ extension Array where Element: NSTextCheckingResult {
+}
+
+extension SequenceType {
+    // swiftlint:disable line_length
```
if you just want to disable for the next line (as opposed to the entire rest of the file, as is happening currently), change this to `// swiftlint:disable:next line_length`

justMaku commented:
@jpsim okay, fixed that tiny thing and also rebased, thanks for all the help!

justMaku commented:
Hey @jpsim, any chance you could take another look on this? :)

marcelofabri commented:
@justMaku Are you sure you've rebased this properly? This is failing on CI because there's no `SequenceType` in Swift 3.0 (it was renamed).

justMaku commented:
Uhm, this is embarrassing 😅. Somehow I was working on a stale master branch still supporting Swift 2.0

jpsim commented:
Thanks for the PR @justMaku! I addressed the merge conflict and refactored things a bit, so I'll now be merging this as part of #945.
